crates/air/src/commands/format.rs

```diff
@@ -1,64 +1,312 @@
+fn judge_entry(entry: DirEntry) -> Option<PathBuf> {
+    // Ignore directories
+    if entry.file_type().map_or(true, |ft| ft.is_dir()) {
+        return None;
+    }
+
+    // Accept all files that are passed-in directly as long as it is an R file
+    if entry.depth() == 0 {
+        let path = entry.into_path();
+
+        if air_fs::has_r_extension(&path) {
+            return Some(path);
+        } else {
+            return None;
+        }
+    }
+
+    // Otherwise check if we should accept this entry
+    // TODO: Many other checks based on user exclude/includes
+    let path = entry.into_path();
+
+    if !air_fs::has_r_extension(&path) {
+        return None;
     }
 
-    // TODO: Respect user specified `LineEnding` option too, not just inferred line endings
-    let line_ending = match line_ending {
+    Some(path)
+}
+
+pub(crate) enum FormatFileResult {
+    Formatted(PathBuf),
+    Unchanged,
+}
+
+impl FormatFileResult {
+    fn is_changed(&self) -> bool {
+        matches!(self, FormatFileResult::Formatted(_))
+    }
+}
+
+// TODO: Take workspace `FormatOptions` that get resolved to `RFormatOptions`
+// for the formatter here. Respect user specified `LineEnding` option too, and
+// only use inferred endings when `FormatOptions::LineEnding::Auto` is used.
+fn format_file(path: PathBuf, mode: FormatMode) -> Result<FormatFileResult, FormatCommandError> {
+    let source = std::fs::read_to_string(&path)
+        .map_err(|err| FormatCommandError::Read(path.clone(), err))?;
+
+    let line_ending = match line_ending::infer(&source) {
         LineEnding::Lf => biome_formatter::LineEnding::Lf,
         LineEnding::Crlf => biome_formatter::LineEnding::Crlf,
     };
+    let options = RFormatOptions::default().with_line_ending(line_ending);
+
+    let source = line_ending::normalize(source);
+    let formatted = match format_source(source.as_str(), options) {
+        Ok(formatted) => formatted,
+        Err(err) => return Err(FormatCommandError::Format(path.clone(), err)),
+    };
+
+    // TODO: We rarely ever take advantage of this optimization on Windows right
+    // now. We always normalize on entry but we apply the requested line ending
+    // on exit (so on Windows we often infer CRLF on entry and normalize to
+    // LF, but apply CRLF on exit so `source` and `new` always have different
+    // line endings). We probably need to compare pre-normalized against
+    // post-formatted output?
+    match formatted {
+        FormattedSource::Formatted(new) => {
+            match mode {
+                FormatMode::Write => {
+                    std::fs::write(&path, new)
+                        .map_err(|err| FormatCommandError::Write(path.clone(), err))?;
+                }
+                FormatMode::Check => {}
+            }
+            Ok(FormatFileResult::Formatted(path))
+        }
+        FormattedSource::Unchanged => Ok(FormatFileResult::Unchanged),
+    }
+}
```

Comment:
`format_source()` takes `&str` and parses and formats it. It returns `FormattedSource` which nicely tells you if the source is `Unchanged` or not. This is super nice as it lets you optimize by avoiding writing back to file when nothing changed. I think the LSP should also use something like this but I had trouble thinking about exactly what to generalize from here for use in the LSP, since the LSP already has its own copy of the CST that it would prefer to pass in.

Reply:
We do have an issue with our string normalization on Windows that prevents this from working well.

- Windows file contains CRLF
- We normalize to LF and store as `source`, but store `LineEnding::CRLF` for the formatter to use
- Formatter emits `formatted` with CRLF endings
- `source` and `formatted` now _always_ look different

I think we should explore not doing the normalization (neither biome nor ruff do it) as I think our tooling can totally handle it, and it would open the door for better optimizations